src/PackageType/BasePackage.php

```diff
@@ -0,0 +1,309 @@
+<?php
+/**
+ * Base package.
+ *
+ * @package SatisPress
+ * @license GPL-2.0-or-later
+ * @since 0.3.0
+ */
+
+declare ( strict_types = 1 );
+
+namespace SatisPress\PackageType;
+
+use SatisPress\Exception\InvalidReleaseVersion;
+use SatisPress\Package;
+use SatisPress\Release;
+
+/**
+ * Base package class.
+ *
+ * @since 0.3.0
+ */
+class BasePackage implements \ArrayAccess, Package {
+	/**
+	 * Package name.
+	 *
+	 * @var string
+	 */
+	protected $name = '';
+
+	/**
+	 * Author.
+	 *
+	 * @var string
+	 */
+	protected $author = '';
+
+	/**
+	 * Author URL.
+	 *
+	 * @var string
+	 */
+	protected $author_url = '';
+
+	/**
+	 * Description.
+	 *
+	 * @var string
+	 */
+	protected $description = '';
+
+	/**
+	 * Homepage URL.
+	 *
+	 * @var string
+	 */
+	protected $homepage = '';
+
+	/**
+	 * Whether the package is installed.
+	 *
+	 * @var boolean
+	 */
+	protected $is_installed = false;
+
+	/**
+	 * Releases.
+	 *
+	 * @var Release[]
+	 */
+	protected $releases = [];
+
+	/**
+	 * Package slug.
+	 *
+	 * @var string
+	 */
+	protected $slug = '';
+
+	/**
+	 * Package type.
+	 *
+	 * @var string
+	 */
+	protected $type = '';
+
+	/**
+	 * Magic setter.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param string $name  Property name.
+	 * @param mixed  $value Property value.
+	 */
+	public function __set( string $name, $value ) {
+		// Don't allow undefined properties to be set.
+		return;
+	}
+
+	/**
+	 * Retrieve the author.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_author(): string {
+		return $this->author;
+	}
+
+	/**
+	 * Retrieve the author URL.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_author_url(): string {
+		return $this->author_url;
+	}
+
+	/**
+	 * Retrieve the description.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_description(): string {
+		return $this->description;
+	}
+
+	/**
+	 * Retrieve the homepage URL.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_homepage(): string {
+		return $this->homepage;
+	}
+
+	/**
+	 * Retrieve the name.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_name(): string {
+		return $this->name;
+	}
+
+	/**
+	 * Retrieve the slug.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_slug(): string {
+		return $this->slug;
+	}
+
+	/**
+	 * Retrieve the package type.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_type(): string {
+		return $this->type;
+	}
+
+	/**
+	 * Whether the package is installed.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return boolean
+	 */
+	public function is_installed(): bool {
+		return $this->is_installed;
+	}
+
+	/**
+	 * Whether the package has any releases.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return boolean
+	 */
+	public function has_releases(): bool {
```

Comment:
The interactions with `$this->releases` is very much treating it like a collection / ArrayObject - checking it's not empty, getting a single one, getting the latest (first) one, etc.

Reply:
It is a lot like a collection and I thought about converting it to one, but would need to make sure it's immutable to mimic the current handling.